src/bmp/decoder.rs

```diff
@@ -0,0 +1,453 @@
+                },
+            };
+            self.has_loaded_metadata = true;
+        }
+        Ok(())
+    }
+
+    fn get_palette_size(&mut self) -> ImageResult<usize> {
+        match self.colors_used {
+            0 => match self.bit_count {
+                8 | 4 | 1 => Ok(1 << self.bit_count),
+                _ => Err(ImageError::FormatError("Invalid bit count for palletized BMP".to_string()))
+            },
+            _ => {
+                if self.colors_used > 1 << self.bit_count {
+                    return Err(ImageError::FormatError(format!(
+                        "Palette size {} exceeds maximum size for BMP with bit count of {}",
+                        self.colors_used, self.bit_count
+                    )))
+                }
+                Ok(self.colors_used as usize)
+            }
+        }
+    }
+
+    fn read_palette(&mut self) -> ImageResult<()> {
+        let bytes_per_color = match self.bmp_header_type { BMPHeaderType::CoreHeader => 3, _ => 4};
+        let palette_size = try!(self.get_palette_size());
+        let length = palette_size * bytes_per_color;
+        let mut buf = Vec::with_capacity(length as usize);
+
+        try!(self.r.by_ref().take(length as u64).read_to_end(&mut buf));
+        let p: Vec<(u8, u8, u8)> = (0usize..palette_size as usize).map(|i| {
+            let b = buf[bytes_per_color * i];
+            let g = buf[bytes_per_color * i + 1];
+            let r = buf[bytes_per_color * i + 2];
+            (r, g, b)
+        }).collect();
+
+        self.palette = Some(p);
+
+        Ok(())
+    }
+
+    fn read_color_index_data(&mut self) -> ImageResult<Vec<u8>> {
+        let row_byte_length = ((self.bit_count as u32 * self.width as u32 + 31) / 32 * 4) as usize;
+        let indexes_per_byte = 8 / self.bit_count;
+        let bit_mask = ((1 << self.bit_count as u16) - 1) as u8;
+        let mut result: Vec<u8> = repeat(0u8).take(self.width as usize * self.height as usize).collect();
+    
+        try!(self.r.seek(SeekFrom::Start(self.data_offset)));
+        for h in 0..self.height {
+            let mut line = Vec::with_capacity(row_byte_length);
+            try!(self.r.by_ref().take(row_byte_length as u64).read_to_end(&mut line));
+
+            let x = if self.top_down { h } else { self.height - h - 1 };
+            let mut y = 0;
+            
+            for i in 0..line.len() {
+                let byte = line[i];
+                for j in 0..indexes_per_byte {
+                    if y >= self.width {
+                        break;
+                    }
+                    result[x as usize * self.width as usize + y as usize] = byte >> (8 - self.bit_count * (j + 1)) & bit_mask;
+                    y += 1;
+                }
+            }
+        }
+
+        Ok(result)
+    }
+
+    fn read_palletized_pixel_data(&mut self) -> ImageResult<Vec<u8>> {
+        let mut pixel_data: Vec<u8> = repeat(0u8).take(self.width as usize * self.height as usize * 3).collect();
```

Comment:
That would be more clear as `let mut pixel_data = vec![0; 3 * self.width as usize * self.height as usize];`